REACT_STYLEGUIDE.MD

```diff
@@ -45,14 +44,31 @@ Bad:
     };
   },
   onClick() {
-    this.setState({ clickCount: this.state.clickCount++ });
+    this.setState(prevState => ({
+      clickCount: prevState.clickCount + 1
+    }));
   },
-  render({ onClick }) {
+  render() {
     return <button className="kuiButton" onClick={this.onClick}/>
   }
 });
 ```
 
+### When a state change only involves the previous state, pass setState a function instead of an object.
```

Comment:
I would favor requiring a function when at least one of the state or the props is used, a function should be passed.

As @chrisronline said, `setState()` not immediately sets the state but schedules a state update to be performed at a convenient time. Maybe we want to also link https://github.com/acdlite/react-fiber-architecture and https://www.youtube.com/watch?v=ZCuYPiUIONs for explanations of how this will be even more important when the fibers-based implementation is released.